src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = {};
+
+  for (const key in state) {
+    if (state.hasOwnProperty(key)) {
+      currentState[key] = state[key];
+    }
+  }
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      for (const key in action.extraData) {
+        if (action.extraData.hasOwnProperty(key)) {
+          currentState[key] = action.extraData[key];
+        }
+      }
+    } else if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    stateHistory.push({ ...currentState });
+  }
+
+  return stateHistory;
 }
```

Comment:
Consider using a `switch` statement instead of multiple `if`/`else if` statements for better readability and maintainability, especially since you're dealing with a limited set of `action.type` values.